### PR TITLE
[4.0] DRAFT static messages for com_redirect/finder etc

### PR DESCRIPTION
**Pull request by brianteeman**
@HLeithner 
as discussed this is a draft of displaying a static message in com_redirects instead of the ever repeating alerts.

I would massively appreciate a code review to see if this is correct (it works but maybe the code can be improved).

We will also need something (javascript?) to reload the page after saving a change in the plugin modal so that the message is refreshed. I assume that this will be something like the code in admin-add_module.js but I really am stabbing in the dark with es6

I did look at various other ways of achieving this including a layout override for messages.php but nothing else worked well and didnt also effect the alerts that we want for item saved etc

The same approach can be used for com_finder

Related discussion at #27380

### CODE REVIEW

**Comment by richard67** [comment]
@brianteeman Shouldn't the `COM_REDIRECT_PLUGIN_MODAL_DISABLED` message be shown in any case if the plugin is disabled, regardless of `collectUrlsEnabled`?

If so, code could be changed and simplified from

```
if (!$pluginEnabled && !$collectUrlsEnabled)
{
	$this->msg = Text::sprintf('COM_REDIRECT_PLUGIN_MODAL_DISABLED', $link);
}
if ($pluginEnabled && !$collectUrlsEnabled)
{
	$this->msg = Text::sprintf('COM_REDIRECT_COLLECT_MODAL_URLS_DISABLED', Text::_('COM_REDIRECT_PLUGIN_ENABLED'), $link);
}
```

to

```
if (!$pluginEnabled)
{
	$this->msg = Text::sprintf('COM_REDIRECT_PLUGIN_MODAL_DISABLED', $link);
}
elseif (!$collectUrlsEnabled)
{
	$this->msg = Text::sprintf('COM_REDIRECT_COLLECT_MODAL_URLS_DISABLED', Text::_('COM_REDIRECT_PLUGIN_ENABLED'), $link);
}
```

**Comment by brianteeman** [comment]
it should be
- if plugin disabled
- if plugin enabled and collect disabled

**Comment by richard67** [comment]
> it should be
> 
>     * if plugin disabled
> 
>     * if plugin enabled and collect disabled

@brianteeman Yes, but like you have it now it is

- if plugin disabled and collect disabled
- if plugin enabled and collect disabled

If you use 2 `if` or one `if` and `else` is then a matter of taste.

But the first if condition `if (!$pluginEnabled && !$collectUrlsEnabled)` for showing the `COM_REDIRECT_PLUGIN_MODAL_DISABLED`text needs to be corrected to `if (!$pluginEnabled)`.

**Comment by richard67** [comment]
@brianteeman Code review looks good to me. Is it still a draft, or shall we start testing?

**Comment by brianteeman** [comment]
you can test but I really want to k ow from @HLeithner if this is a valid approach
and still need some help with js part

**Comment by richard67** [comment]
Ah, yes, I forgot about the js part. Sorry I can't help with that.

**Comment by Quy** [comment]
Richard’s proposed code is better in that when the first check is true then there is no need to perform the second check.

**Comment by richard67** [comment]
Well, my code maybe might save a few atoseconds, but maybe Brian's code is easier to understand when reading, so I am ok with it.

File: administrator/components/com_redirect/tmpl/links/default.php [hunk 1]
```diff
@@ -14,17 +14,45 @@
...
+$pluginEnabled      = PluginHelper::isEnabled('system', 'redirect');
+$collectUrlsEnabled = RedirectHelper::collectUrlsEnabled();
+$redirectPluginId   = RedirectHelper::getRedirectPluginId();
+$link = HTMLHelper::_(
+	'link',
+	'#plugin' . $redirectPluginId . 'Modal',
+	Text::_('COM_REDIRECT_SYSTEM_PLUGIN'),
+	'class="alert-link" data-toggle="modal" id="title-' . $redirectPluginId . '"'
+);
+if (!$pluginEnabled)
```
Comment (Quy) [inline administrator/components/com_redirect/tmpl/links/default.php]:
```suggestion
if (!$pluginEnabled)
```

**Comment by HLeithner** [comment]
There is way too much logic in the template, that should all be done in the model and view in my opinion and only a "if $showwarning then echo" should be in the template...

**Comment by brianteeman** [comment]
I thought that would be the case. Any help greatly appreciated

**Comment by HLeithner** [comment]
I will see what I can do tomorrow

**Comment by infograf768** [comment]
> There is way too much logic in the template, that should all be done in the model and view in my opinion and only a "if $showwarning then echo" should be in the template...

That is exactly what I suggested in https://github.com/joomla/joomla-cms/pull/27380#issuecomment-569718041

1. Create a new method in the model
2. Set the variable obtained in the `display` method in HtmlView
3. And only the div and echo in default.php if variable is not null

**Comment by richard67** [comment]
PHP code review looks good to me, and I've just tested it with success. Now only JS part is missing. Maybe this could also update the redirect view after having enabled the plugin, so the displayed messages adapt accordingly.

P.S.: Maybe just force a page reload after having enabled the plugin?

**Comment by brianteeman** [comment]
Now added the same for com_finder with the added bonus that the plugin settings are now changed in a modal and not by redirevting to com_plugins

File: administrator/language/en-GB/com_redirect.ini [hunk 2]
```diff
@@ -12,7 +12,7 @@ COM_REDIRECT_BULK_SEPARATOR_LABEL="Bulk Import Separator"
 COM_REDIRECT_BUTTON_UPDATE_LINKS="Update Links"
 COM_REDIRECT_CLEAR_FAIL="Failed to delete disabled links."
 COM_REDIRECT_CLEAR_SUCCESS="All disabled links have been deleted."
-COM_REDIRECT_COLLECT_MODAL_URLS_DISABLED="%1$s The 'Collect URLs' option in the %2$s is disabled. Error page URLs will not be collected by this component."
+COM_REDIRECT_COLLECT_MODAL_URLS_DISABLED="The 'Collect URLs' option in the %1$s is disabled. Error page URLs will not be collected by this component."
 COM_REDIRECT_COLLECT_URLS_ENABLED="%1$s The option 'Collect URLs' is enabled."
```
Comment (Quy) [inline administrator/language/en-GB/com_redirect.ini]:
```suggestion
```

File: administrator/components/com_redirect/tmpl/links/default.php [hunk 3]
```diff
@@ -22,9 +22,18 @@
...
 ?>
+
 <form action="<?php echo Route::_('index.php?option=com_redirect&view=links'); ?>" method="post" name="adminForm" id="adminForm">
 	<div id="j-main-container" class="j-main-container">
 		<?php echo LayoutHelper::render('joomla.searchtools.default', array('view' => $this)); ?>
+		<?php if (!empty($this->informationMessages)) : ?>
+		<?php foreach($this->informationMessages as $message): ?>
+			<div class="alert alert-info" aria-live="polite">
+				<span class="fa fa-info-circle" aria-hidden="true"></span><span class="sr-only"><?php echo Text::_('INFO'); ?></span>
+				<?php echo Text::sprintf(...$message['message']); ?>
```
Comment (Quy) [inline administrator/components/com_redirect/tmpl/links/default.php:33]:
```suggestion
				<?php echo Text::sprintf($message['message']); ?>
```

File: administrator/components/com_redirect/tmpl/links/default.php [hunk 4]
```diff
@@ -22,9 +22,18 @@
 $listOrder = $this->escape($this->state->get('list.ordering'));
 $listDirn  = $this->escape($this->state->get('list.direction'));
 ?>
+
 <form action="<?php echo Route::_('index.php?option=com_redirect&view=links'); ?>" method="post" name="adminForm" id="adminForm">
 	<div id="j-main-container" class="j-main-container">
 		<?php echo LayoutHelper::render('joomla.searchtools.default', array('view' => $this)); ?>
+		<?php if (!empty($this->informationMessages)) : ?>
+		<?php foreach($this->informationMessages as $message): ?>
```
Comment (Quy) [inline administrator/components/com_redirect/tmpl/links/default.php]:
```suggestion
		<?php foreach($this->informationMessages as $message) : ?>
```

File: administrator/components/com_redirect/tmpl/links/default.php (hunk 3 above)
Comment (brianteeman) [reply administrator/components/com_redirect/tmpl/links/default.php:33]: Please check this with @HLeithner

**Comment by brianteeman** [comment]
@fedik any help on the javascript?

**Comment by Fedik** [comment]
@brianteeman I have made pull request, please check

Note, I would prefer inline editing, without modal. Modal always hard on mobiles.

Like:
Open plugin form => save&close => redirect back to com_finder,
In theory that should be possible with append of `&return=base64-url-to-return`, and maybe with some editing of the plugin form layout .

**Comment by brianteeman** [comment]
@fedik maybe i did something wrong or maybe i explained the problem wrong

with content-smart search plugin disabled you get the message to enable the plugin
you click on the plugin and enable it and save and close

at this point I was expecting the page to refresh so that the message is no longer displayed

**Comment by Fedik** [comment]
Did you get the page reload after "save and close"? (even if message still there)

After JS changes, it should be like you said (not forget npm , to update JS).
One thing I am afraid, it can show cached page, by browser.

**Comment by brianteeman** [comment]
> Did you get the page reload after "save and close"? (even if message still there)

No

**Comment by brianteeman** [comment]
Closing as we no longer have popup alerts

**Pull request closed without merging**